Story/1824 Mobile UI: Improve Odoo printing logic

jrbirch99 opened the pull request:
The print logic in odoo_print has been changed such that the missing report types are reported on properly. Unit tests were adjusted for this change. 

I also spotted that spool_report() would output the wrong thing if fed multiple reports of the same report_type. So I fixed it, and added a unit test for it. 

I also refactored the spool_report() method to make it easier to follow.

jrbirch99 committed:
[FIX] print: Fix the logic in the odoo print module.

- Refactor the spool_report() method; split it up into smaller methods that are easier to read.
- Ensure that the missing report types are reported on properly such that the logic follows: report types found - report types needed
- During spool_test_page() only call the report test page corresponding to the printer.report_type. This has meant changing some of the data too.

Story/1824

Signed-off-by: Jack Birch <jack.birch@unipart.io>

jrbirch99 committed:
[FIX] print: Allow multiple reports of the same report type to be printed at the same time.

If sent multiple reports of the same report type only the last report was printed. This issue was that the value in the {key : value} pair of the dictionary was being overwritten instead of extended.

A unit test was also added for this edge case.

Story/1824

Signed-off-by: Jack Birch <jack.birch@unipart.io>

jrbirch99 committed:
[IMP] print: Formatting changes.

Story/1824

Signed-off-by: Jack Birch <jack.birch@unipart.io>

caleb-shelton commented on addons/print/models/print_printer.py:
```diff
@@ -100,7 +107,14 @@ def set_default_printer(self):
     def printers(self, report_type=None, raise_if_not_found=False):
-        """Determine printers to use"""
+        """
+        Purpose: Determine printers to use for a given report type. If no printers are found
```
nit: remove `Purpose` as it is already implied by writing a doc string. But also I don't think it is necessary to include the return type in the docstring as we don't tend to do this elsewhere.

It would be nice to have some clearer guidelines for docstrings though so we can standardise them everywhere :)

jrbirch99 replied:
Yeah I think clearer guidelines are needed for sure. I want to keep the return type just because it makes it clear which data type is being returned, I think if we were working with a statically typed language there would be no need (because we would already specify it) but because python is dynamically typed it can take on any type, so best to keep track of what we think it is. I know there was a story recently where a method was returning a different data type to what we expected so I think it might be better to keep it in. 

Personally I like having the "Purpose" as it makes the doc-string clear and differentiates that between the returns :smile:

caleb-shelton replied:
@ws-mps do we have any guidelines surrounding how docstrings should be written?

lcucurachi replied:
I think having the return is fine, it's actually a requirement to have one for some docstring standards/formats. I do agree with Caleb on the purpose though, not really necessary to have to say that you are explaining the purpose since the docstring is documentation of a function. I'd say drop the "purpose" on the docstrings, no need for that. It's easy to get lost on these but realistically docstrings are already explaining a function.

lcucurachi replied:
> @ws-mps do we have any guidelines surrounding how docstrings should be written?

We need to adopt a standard, we haven't done it yet (I believe). I have done some research on this which would be helpful to decide which one to use. Search this page on the wiki (StoryX1871)

jrbirch99 replied:
So I have removed the purpose but I want to keep the returns :smile:

lcucurachi commented on addons/print/models/print_printer.py:
```diff
@@ -195,70 +251,66 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
             missing_labels = [self.get_report_type_label(rt) for rt in missing]
             raise UserError(_("Missing reports of types: %s") % ", ".join(missing_labels))
 
-        # CPCL reports require number of copies passed into the template via data
```
Might be nice to keep this comment where you placed the code below so that we don't forget why we do this?

jrbirch99 replied:
This one I did keep, it should be on line 214

armandcela commented on addons/print/models/print_printer.py:
```diff
@@ -177,14 +191,56 @@ def spool(self, document, title=None, copies=1):
             return True
 
-        # Identify reports
+        reports = self._fetch_reports_for_printing(report_name)
```
Nit: This is personal in fact, i dont like the empty lines on every method. Maybe a comment if the name is not explanatory. To me empty lines needed when we are writing a block of logic.

jrbirch99 replied:
changed

lcucurachi reviewed (commented):
LGTM, have you also done some manual tests trying to print something to see that it still works?

PeterC10 commented on addons/print/models/print_printer.py:
```diff
@@ -32,10 +33,16 @@ class Printer(models.Model):
-        string="Full Name", compute="_compute_full_name", store=True, 
+        string="Full Name",
+        compute="_compute_full_name",
+        store=True,
```
Nit: Doesn't matter but I think black is splitting these out onto multiple lines because there is a comma after the last element, could you remove these and run black again? `barcode` and `queue` at least should be 1 line!

PeterC10 commented on addons/print/models/print_printer.py:
```diff
@@ -177,14 +191,53 @@ def spool(self, document, title=None, copies=1):
+        for report in reports:
+            documents[report.report_type].append(
+                (
+                    ("%s %s" % (report.name, str(docids))) if title is None else title,
```
Suggestion: Put these two values into variables, e.g. `doc_title` and `rendered_doc`

These few lines are tricky to decipher atm.

Something like this is easier to read imo.

```
doc_title = title or ("%s %s" % (report.name, str(docids)))
rendered_doc = report._render(docids, cpcl_data if report.report_type == "qweb-cpcl" else data)[0]

documents[report.report_type].append((doc_title, rendered_doc))
```

PeterC10 commented on addons/print/models/print_printer.py:
```diff
@@ -195,70 +248,66 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
+        report type.
+        Return: dict{ ir.action.report() : print.printer() }
+        """
+        printer_raise_if_not_found = len(required_report_types) == 1
```
Q: Do we only raise an error for printer not being found if there are multiple report types because we might have be able to print out some of the types but not others?

Might be worth adding a comment or something in the doc-string.

jrbirch99 replied:
On closer inspection I realised that if we always pass `True` then` _check_every_report_type_has_printer()` is redundant. 

`_check_every_report_type_has_printer()` would check that each report type has a printer, but if we just raise the error when no printer is found for a report type - which we can do in `printers()` - we do not need `_check_every_report_type_has_printer()`.

jrbirch99 committed:
[IMP] print: Refactor the spool_report() method.

If a printer can not be found for a report a UserError is raised, which will inform that a printer for said report is missing, consequentally the report cannot be printed.

_check_every_report_type_has_printer() was found to be duplicated functionality as it raises an error when there is no printer for a report type, but an error is already raised in printers() if a printer can't be found for the desired report type.

Story/1824

Signed-off-by: Jack Birch <jack.birch@unipart.io>

PeterC10 commented on addons/print/models/print_printer.py:
```diff
@@ -216,14 +208,11 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
-                )
-            )
+            doc_title = title or ("%s %s" % (report.name, str(docids)))
+            rendered_doc = report._render(
```
Sorry to be awkward, but this is still a bit awkward to read, black doesn't help here.

Could we add `report_data`?

So:

```
report_data = cpcl_data if report.report_type == "qweb-cpcl" else data
rendered_doc = report._render(docids, report_data)[0]
```

Again, just a suggestion. Happy for this to be resolved.
